README.md
# Free Icons

This crate will generate helper data structures for accessing icons. It is highly inspired by [dioxus-free-icons](https://github.com/nissy-dev/dioxus-free-icons). Because I just need pure SVG icons to be embedded in HTML without any client-side javascript, I decided to write this crate.

## Usage

If you need raw SVG icons without any modification, you can use the general `get` function:

```rust
// generate a heroicon
let icon = free_icons::get(IconType::Heroicons(Heroicons::Outline), "academic-cap");
assert_eq!(icon, r###"<svg viewbox="0 0 24 24" fill=none height=24 width=24 xmlns=http://www.w3.org/2000/svg><path d="M11.6998 2.80541C11.8912 2.72176 12.1089 2.72176 12.3003 2.80541C16.0192 4.43023 19.5437 6.41649 22.8295 8.71968C23.0673 8.88636 23.1875 9.17532 23.1381 9.46147C23.0887 9.74762 22.8785 9.97953 22.5986 10.0568C21.9137 10.2459 21.2347 10.4495 20.5618 10.6665C17.8307 11.5473 15.2018 12.6555 12.6972 13.9689L12.6939 13.9706C12.5803 14.0302 12.467 14.0902 12.354 14.1506C12.1331 14.2686 11.8679 14.2686 11.6471 14.1506C11.533 14.0896 11.4186 14.029 11.3039 13.9689C10.0655 13.3195 8.79658 12.7202 7.5 12.1738V11.9501C7.5 11.8187 7.56742 11.7022 7.67173 11.639C9.17685 10.7271 10.7294 9.88577 12.3247 9.11948C12.6981 8.94014 12.8554 8.49207 12.6761 8.1187C12.4967 7.74532 12.0486 7.58803 11.6753 7.76737C10.036 8.55475 8.44086 9.41921 6.89449 10.356C6.44111 10.6307 6.13632 11.0803 6.03607 11.5839C5.18115 11.255 4.31499 10.9487 3.43829 10.666C2.76546 10.449 2.08644 10.2458 1.40154 10.0568C1.12162 9.97952 0.911461 9.74761 0.86204 9.46146C0.812619 9.17531 0.932824 8.88634 1.17061 8.71967C4.45645 6.41648 7.98097 4.43023 11.6998 2.80541Z" fill=#0F172A /><path d="M13.0609 15.4735C15.4997 14.1704 18.0621 13.0688 20.7258 12.1907C20.8601 13.6055 20.9458 15.0344 20.9813 16.4756C20.9889 16.7848 20.8059 17.067 20.5205 17.1862C17.6693 18.3765 14.9574 19.8342 12.4159 21.5278C12.1641 21.6957 11.836 21.6957 11.5841 21.5278C9.04267 19.8342 6.33073 18.3765 3.4796 17.1862C3.19416 17.067 3.01116 16.7848 3.01878 16.4756C3.05429 15.0343 3.14001 13.6053 3.27427 12.1904C4.19527 12.494 5.10415 12.8243 6 13.1804V14.4508C5.55165 14.7102 5.25 15.1949 5.25 15.7501C5.25 16.2454 5.49008 16.6847 5.86022 16.9578C5.7707 17.3385 5.63822 17.7109 5.46277 18.0676C5.91546 18.2813 6.36429 18.5018 6.8091 18.7291C7.06243 18.2138 7.24612 17.673 7.36014 17.1208C7.88449 16.8871 8.25 16.3613 8.25 15.7501C8.25 15.1949 7.94835 14.7102 7.5 14.4508V13.806C8.6714 14.3178 9.81885 14.8744 10.9402 15.4735C11.6028 15.8276 12.3983 15.8276 13.0609 15.4735Z" fill=#0F172A /><path d="M4.46222 19.4624C4.88136 19.0433 5.21502 18.5712 5.46277 18.0676C5.91546 18.2813 6.36429 18.5018 6.8091 18.7291C6.49055 19.3769 6.06164 19.9843 5.52288 20.5231C5.22999 20.816 4.75512 20.816 4.46222 20.5231C4.16933 20.2302 4.16933 19.7553 4.46222 19.4624Z" fill=#0F172A /></svg>"###);
```

If you want to add extra attributes to the SVG icon, you could use the specific function (currently bootstrap, feather, font_awesome, heroicons, ionicons, octicons are supported):

```rust

```rust
// generate a bootstrap alarm icon
let attrs = IconAttrs::default()
    .class("h-8 w-8 text-white")
    .fill("none")
    .stroke_color("currentColor");
let icon = free_icons::bootstrap("alarm", true, &attrs);
assert_eq!(icon, r###"<svg class="h-8 w-8 text-white" fill="none" stroke="currentColor" viewbox="0 0 16 16" fill=currentColor height=16 width=16 xmlns=http://www.w3.org/2000/svg><path d="M6 .5a.5.5 0 0 1 .5-.5h3a.5.5 0 0 1 0 1H9v1.07a7.001 7.001 0 0 1 3.274 12.474l.601.602a.5.5 0 0 1-.707.708l-.746-.746A6.97 6.97 0 0 1 8 16a6.97 6.97 0 0 1-3.422-.892l-.746.746a.5.5 0 0 1-.707-.708l.602-.602A7.001 7.001 0 0 1 7 2.07V1h-.5A.5.5 0 0 1 6 .5zm2.5 5a.5.5 0 0 0-1 0v3.362l-1.429 2.38a.5.5 0 1 0 .858.515l1.5-2.5A.5.5 0 0 0 8.5 9V5.5zM.86 5.387A2.5 2.5 0 1 1 4.387 1.86 8.035 8.035 0 0 0 .86 5.387zM11.613 1.86a2.5 2.5 0 1 1 3.527 3.527 8.035 8.035 0 0 0-3.527-3.527z"/></svg>"###);
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [conventional commits](https://www.conventionalcommits.org/) for commit guidelines.

---
## [0.7.0](https://github.com/tyrchen/free-icons/compare/v0.6.0..v0.7.0) - 2023-12-27

### Features

- use latest html_minify - ([a077203](https://github.com/tyrchen/free-icons/commit/a0772036e5d1cfe3a7e1d7d9c5a6bc4f79975026)) - Tyr Chen

### Miscellaneous Chores

- updte dependencies - ([73b6cfa](https://github.com/tyrchen/free-icons/commit/73b6cfac36362597961afbda2ce6b9207facf228)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([9562046](https://github.com/tyrchen/free-icons/commit/956204601e9976b620e42b7c24a0cdf78a9582c7)) - Tyr Chen

---
## [0.6.0](https://github.com/tyrchen/free-icons/compare/v0.5.5..v0.6.0) - 2023-12-18

### Features

- update submodules - ([3e51de5](https://github.com/tyrchen/free-icons/commit/3e51de59c9616433c7bf46381e716c72d6f786c5)) - Tyr Chen

### Miscellaneous Chores

- update dependencies - ([e3b6e18](https://github.com/tyrchen/free-icons/commit/e3b6e188bc70ef82053444dcbac2b026d6cb87c8)) - Tyr Chen
- update gh action and cliff - ([c295aed](https://github.com/tyrchen/free-icons/commit/c295aed7eba7f09fab931127c0838e4419e444cd)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([a1af1da](https://github.com/tyrchen/free-icons/commit/a1af1da6ba8a614b2b48906894984ce7faf81187)) - Tyr Chen

---
## [0.5.5](https://github.com/tyrchen/free-icons/compare/v0.5.4..v0.5.5) - 2023-02-14

### Miscellaneous Chores

- bump deps - ([cbb99e2](https://github.com/tyrchen/free-icons/commit/cbb99e2b10f36ae91ddb5ee2caa397c84b011b16)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([2535c11](https://github.com/tyrchen/free-icons/commit/2535c115bee95868cd96fdda700c473e6068d9c7)) - Tyr Chen

---
## [0.5.4](https://github.com/tyrchen/free-icons/compare/v0.5.3..v0.5.4) - 2023-02-01

### Features

- merge the brand icons back to regular/solid categories for font-awesome - ([f473c8a](https://github.com/tyrchen/free-icons/commit/f473c8a500de8650bcd917e9636788a70cbd861e)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([6ef7ff7](https://github.com/tyrchen/free-icons/commit/6ef7ff7c74fffa754006bb8d7640d0b5b519c4ae)) - Tyr Chen

---
## [0.5.3](https://github.com/tyrchen/free-icons/compare/v0.5.2..v0.5.3) - 2023-01-20

### Miscellaneous Chores

- bump minijinja to 0.30.0 - ([236511b](https://github.com/tyrchen/free-icons/commit/236511b7883e64eac4ea6ebd9021a41d52446195)) - Tyr Chen

---
## [0.5.2](https://github.com/tyrchen/free-icons/compare/v0.5.1..v0.5.2) - 2023-01-11

### Other

- Update CHANGELOG.md - ([0e38771](https://github.com/tyrchen/free-icons/commit/0e38771f38595a4a5b52f99cdb2572efd055f3ab)) - Tyr Chen
- update deps - ([059a115](https://github.com/tyrchen/free-icons/commit/059a115ac4fe5ebbed4032051ab7f2f325b52b3c)) - Tyr Chen

---
## [0.5.1](https://github.com/tyrchen/free-icons/compare/v0.5.0..v0.5.1) - 2022-12-30

### Bug Fixes

- fix cargo doc and upgrade doc - ([1661e27](https://github.com/tyrchen/free-icons/commit/1661e2700f2d861bd6d27e7571be7702bfa83a44)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([3b02bc8](https://github.com/tyrchen/free-icons/commit/3b02bc8a130ec35289eb3a30e683be9d8b04300d)) - Tyr Chen

---
## [0.5.0](https://github.com/tyrchen/free-icons/compare/v0.4.1..v0.5.0) - 2022-12-14

### Features

- upgrade bincode to 2 - ([02a9594](https://github.com/tyrchen/free-icons/commit/02a95943675187f4f1682e4e66fd935b37d9c686)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([75ca6c7](https://github.com/tyrchen/free-icons/commit/75ca6c76531abf3b54ace1b708284a58dd318375)) - Tyr Chen

---
## [0.4.1](https://github.com/tyrchen/free-icons/compare/v0.4.0..v0.4.1) - 2022-12-14

### Other

- Update CHANGELOG.md - ([9423fc7](https://github.com/tyrchen/free-icons/commit/9423fc7f4c397ceb2da40c73df39aad299dd7a39)) - Tyr Chen

### Refactoring

- use minijinjia to replace tera - ([bc12102](https://github.com/tyrchen/free-icons/commit/bc1210286c496050b928db0fe0bca23bfae1738b)) - Tyr Chen

---
## [0.4.0](https://github.com/tyrchen/free-icons/compare/v0.3.0..v0.4.0) - 2022-12-03

### Features

- extend the interface so that it can add extra attributes - ([a3155c9](https://github.com/tyrchen/free-icons/commit/a3155c9d2c6696772497dda081721796e7adb69c)) - Tyr Chen
- improve attrs data structure and fix heroicons (use optimized rather than src) - ([d9e9036](https://github.com/tyrchen/free-icons/commit/d9e9036a567cf8389cd5d8e9d61dd1fff403785e)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([1e29270](https://github.com/tyrchen/free-icons/commit/1e292707d3858ecfb6f08392135ab5f381fe2405)) - Tyr Chen

---
## [0.3.0](https://github.com/tyrchen/free-icons/compare/v0.2.0..v0.3.0) - 2022-12-03

### Features

- allow adding extra class - ([40a82f8](https://github.com/tyrchen/free-icons/commit/40a82f87a01ceee1f5ca776521ff08529f1d2540)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([76bcc32](https://github.com/tyrchen/free-icons/commit/76bcc32ec3589150e5d6b0d26b16c6f22ea75bfe)) - Tyr Chen

---
## [0.2.0](https://github.com/tyrchen/free-icons/compare/v0.1.0..v0.2.0) - 2022-12-03

### Bug Fixes

- fix github action - ([97000e9](https://github.com/tyrchen/free-icons/commit/97000e95b98d38e2d75310f7ec06ded8f3d54b0f)) - Tyr Chen

### Features

- use lazy_static and flate2 to reduce binary size (with the cost of slighter longer loading time) - ([4ee5abc](https://github.com/tyrchen/free-icons/commit/4ee5abc646c6898530634c4563521c61feea16b3)) - Tyr Chen

### Other

- Update CHANGELOG.md - ([02f8d9f](https://github.com/tyrchen/free-icons/commit/02f8d9fc751437b6c6c54fe6d37aacc18310c0ab)) - Tyr Chen

---
## [0.1.0] - 2022-12-03

### Features

- add font-awesome - ([3336905](https://github.com/tyrchen/free-icons/commit/33369050e62b090800b4341bbe48a033f091b856)) - Tyr Chen
- add heroicons - ([87a26af](https://github.com/tyrchen/free-icons/commit/87a26af33be7a5ed55c15d5d5b008c8fba4bbe99)) - Tyr Chen

src/lib.rs
use flate2::bufread::GzDecoder;
use std::{borrow::Cow, collections::HashMap, io::Read};

mod gen;

const MAX_ATTRS: usize = 16;
#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct IconAttrs<'a> {
    data: [(&'a str, Cow<'a, str>); MAX_ATTRS],
    pos: u8,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum IconType {
    #[cfg(feature = "bootstrap")]
    Bootstrap(Bootstrap),
    #[cfg(feature = "feather")]
    Feather(Feather),
    #[cfg(feature = "font-awesome")]
    FontAwesome(FontAwesome),
    #[cfg(feature = "heroicons")]
    Heroicons(Heroicons),
    #[cfg(feature = "ionicons")]
    Ionicons(Ionicons),
    #[cfg(feature = "octicons")]
    Octicons(Octicons),
}

#[cfg(feature = "bootstrap")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Bootstrap {
    Fill,
    Normal,
}

#[cfg(feature = "feather")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Feather {
    Normal,
}

#[cfg(feature = "font-awesome")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FontAwesome {
    Regular,
    Solid,
}

#[cfg(feature = "heroicons")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Heroicons {
    Outline,
    Solid,
}

#[cfg(feature = "ionicons")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Ionicons {
    Outline,
    Sharp,
    Normal,
}

#[cfg(feature = "octicons")]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Octicons {
    Normal,
}

/// retrieve the SVG from incon_type and name
pub fn get(icon_type: IconType, name: &str) -> Option<&'static String> {
    match icon_type {
        #[cfg(feature = "bootstrap")]
        IconType::Bootstrap(icon_type) => match icon_type {
            Bootstrap::Fill => gen::bootstrap::FILL.get(name),
            Bootstrap::Normal => gen::bootstrap::NORMAL.get(name),
        },
        #[cfg(feature = "feather")]
        IconType::Feather(icon_type) => match icon_type {
            Feather::Normal => gen::feather::NORMAL.get(name),
        },
        #[cfg(feature = "font-awesome")]
        IconType::FontAwesome(icon_type) => match icon_type {
            FontAwesome::Regular => gen::font_awesome::REGULAR.get(name),
            FontAwesome::Solid => gen::font_awesome::SOLID.get(name),
        },
        #[cfg(feature = "heroicons")]
        IconType::Heroicons(icon_type) => match icon_type {
            Heroicons::Outline => gen::heroicons::OUTLINE.get(name),
            Heroicons::Solid => gen::heroicons::SOLID.get(name),
        },
        #[cfg(feature = "ionicons")]
        IconType::Ionicons(icon_type) => match icon_type {
            Ionicons::Outline => gen::ionicons::OUTLINE.get(name),
            Ionicons::Sharp => gen::ionicons::SHARP.get(name),
            Ionicons::Normal => gen::ionicons::NORMAL.get(name),
        },
        #[cfg(feature = "octicons")]
        IconType::Octicons(icon_type) => match icon_type {
            Octicons::Normal => gen::octicons::NORMAL.get(name),
        },
    }
}

#[cfg(feature = "bootstrap")]
#[inline(always)]
pub fn bootstrap(name: &str, filled: bool, attrs: IconAttrs) -> Option<String> {
    let svg = if filled {
        gen::bootstrap::FILL.get(name)
    } else {
        gen::bootstrap::NORMAL.get(name)
    };
    attrs.add_to_svg(svg)
}

#[cfg(feature = "feather")]
#[inline(always)]
pub fn feather(name: &str, attrs: IconAttrs) -> Option<String> {
    let svg = gen::feather::NORMAL.get(name);
    attrs.add_to_svg(svg)
}

#[cfg(feature = "font-awesome")]
#[inline(always)]
pub fn font_awesome(name: &str, category: FontAwesome, attrs: IconAttrs) -> Option<String> {
    let svg = match category {
        FontAwesome::Regular => gen::font_awesome::REGULAR.get(name),
        FontAwesome::Solid => gen::font_awesome::SOLID.get(name),
    };
    attrs.add_to_svg(svg)
}

#[cfg(feature = "heroicons")]
#[inline(always)]
pub fn heroicons(name: &str, outline: bool, attrs: IconAttrs) -> Option<String> {
    let svg = if outline {
        gen::heroicons::OUTLINE.get(name)
    } else {
        gen::heroicons::SOLID.get(name)
    };
    attrs.add_to_svg(svg)
}

#[cfg(feature = "ionicons")]
#[inline(always)]
pub fn ionicons(name: &str, category: Ionicons, attrs: IconAttrs) -> Option<String> {
    let svg = match category {
        Ionicons::Outline => gen::ionicons::OUTLINE.get(name),
        Ionicons::Sharp => gen::ionicons::SHARP.get(name),
        Ionicons::Normal => gen::ionicons::NORMAL.get(name),
    };
    attrs.add_to_svg(svg)
}

#[cfg(feature = "octicons")]
#[inline(always)]
pub fn octicons(name: &str, attrs: IconAttrs) -> Option<String> {
    let svg = gen::octicons::NORMAL.get(name);
    attrs.add_to_svg(svg)
}

pub(crate) fn decap(bytes: &[u8]) -> HashMap<String, HashMap<String, String>> {
    let mut gz = GzDecoder::new(bytes);
    let mut uncompressed = Vec::new();
    gz.read_to_end(&mut uncompressed).expect("should decap");
    let (ret, _) = bincode::decode_from_slice(&uncompressed, bincode::config::standard())
        .expect("should deserialize");
    ret
}

impl<'a> IconAttrs<'a> {
    #[inline(always)]
    pub fn class(self, class: &'a str) -> Self {
        self.with("class", class)
    }

    #[inline(always)]
    pub fn fill(self, fill: &'a str) -> Self {
        self.with("fill", fill)
    }

    #[inline(always)]
    pub fn stroke_color(self, stroke_color: &'a str) -> Self {
        self.with("stroke", stroke_color)
    }

    #[inline(always)]
    pub fn stroke_width(self, stroke_width: &'a str) -> Self {
        self.with("stroke-width", stroke_width)
    }

    #[inline(always)]
    pub fn with(self, attr: &'a str, value: &'a str) -> Self {
        let mut data = self.data;
        let mut pos = self.pos;
        data[pos as usize] = (attr, value.into());
        pos = (pos + 1) % MAX_ATTRS as u8;

        Self { data, pos }
    }
    fn add_to_svg(&self, svg: Option<&String>) -> Option<String> {
        if let Some(svg) = svg {
            let mut svg = svg.to_owned();
            let mut attrs = String::new();
            for i in 0..self.pos {
                let (k, v) = &self.data[i as usize];
                attrs.push_str(&format!(" {k}=\"{v}\""));
            }

            if !attrs.is_empty() {
                svg.insert_str(4, &attrs);
            }
            Some(svg)
        } else {
            None
        }
    }
}

#[cfg(feature = "json")]
impl<'a> From<&'a serde_json::Value> for IconAttrs<'a> {
    fn from(value: &'a serde_json::Value) -> Self {
        let mut attrs = IconAttrs::default();
        if let serde_json::Value::Object(map) = value {
            for (k, v) in map {
                let s = match v {
                    serde_json::Value::String(s) => s.into(),
                    serde_json::Value::Number(n) => n.to_string().into(),
                    serde_json::Value::Bool(b) => b.to_string().into(),
                    _ => continue,
                };

                attrs.data[attrs.pos as usize] = (k, s);
                attrs.pos = (attrs.pos + 1) % MAX_ATTRS as u8;
            }
        }
        attrs
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use minify_html::{minify, Cfg};
    const CFG: Cfg = Cfg {
        keep_closing_tags: true,
        do_not_minify_doctype: false,
        ensure_spec_compliant_unquoted_attribute_values: true,
        keep_html_and_head_opening_tags: true,
        keep_spaces_between_attributes: true,
        keep_comments: false,
        minify_css: true,
        minify_js: true,
        remove_bangs: false,
        remove_processing_instructions: false,
        keep_input_type_text_attr: true,
        keep_ssi_comments: false,
        preserve_brace_template_syntax: true,
        preserve_chevron_percent_template_syntax: false,
    };

    #[cfg(feature = "bootstrap")]
    #[test]
    fn bootstrap_icon_fill_should_work() {
        assert_eq!(
            get(IconType::Bootstrap(Bootstrap::Fill), "alarm"),
            Some(&expected(include_str!(
                "../icon_resources/bootstrap/icons/alarm-fill.svg"
            )))
        );
    }

    #[cfg(feature = "bootstrap")]
    #[test]
    fn bootstrap_icon_should_work() {
        assert_eq!(
            get(IconType::Bootstrap(Bootstrap::Normal), "alarm"),
            Some(&expected(include_str!(
                "../icon_resources/bootstrap/icons/alarm.svg"
            )))
        );
    }

    #[cfg(feature = "feather")]
    #[test]
    fn feather_icon_should_work() {
        assert_eq!(
            get(IconType::Feather(Feather::Normal), "activity"),
            Some(&expected(include_str!(
                "../icon_resources/feather/icons/activity.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_icon_brands_should_work() {
        assert_eq!(
            get(IconType::FontAwesome(FontAwesome::Regular), "500px"),
            Some(&expected(include_str!(
                "../icon_resources/font-awesome/svgs/brands/500px.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_icon_regular_should_work() {
        assert_eq!(
            get(IconType::FontAwesome(FontAwesome::Regular), "address-book"),
            Some(&expected(include_str!(
                "../icon_resources/font-awesome/svgs/regular/address-book.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_icon_solid_should_work() {
        assert_eq!(
            get(IconType::FontAwesome(FontAwesome::Solid), "address-book"),
            Some(&expected(include_str!(
                "../icon_resources/font-awesome/svgs/solid/address-book.svg"
            )))
        );
    }

    #[cfg(feature = "heroicons")]
    #[test]
    fn heroicons_icon_outline_should_work() {
        assert_eq!(
            get(IconType::Heroicons(Heroicons::Outline), "academic-cap"),
            Some(&expected(include_str!(
                "../icon_resources/heroicons/optimized/24/outline/academic-cap.svg"
            )))
        );
    }

    #[cfg(feature = "heroicons")]
    #[test]
    fn heroicons_icon_solid_should_work() {
        assert_eq!(
            get(IconType::Heroicons(Heroicons::Solid), "academic-cap"),
            Some(&expected(include_str!(
                "../icon_resources/heroicons/optimized/24/solid/academic-cap.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_icon_outline_should_work() {
        assert_eq!(
            get(IconType::Ionicons(Ionicons::Outline), "alarm"),
            Some(&expected(include_str!(
                "../icon_resources/ionicons/src/svg/alarm-outline.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_icon_sharp_should_work() {
        assert_eq!(
            get(IconType::Ionicons(Ionicons::Sharp), "alarm"),
            Some(&expected(include_str!(
                "../icon_resources/ionicons/src/svg/alarm-sharp.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_icon_should_work() {
        assert_eq!(
            get(IconType::Ionicons(Ionicons::Normal), "alarm"),
            Some(&expected(include_str!(
                "../icon_resources/ionicons/src/svg/alarm.svg"
            )))
        );
    }

    #[cfg(feature = "octicons")]
    #[test]
    fn octicons_icon_should_work() {
        assert_eq!(
            get(IconType::Octicons(Octicons::Normal), "alert"),
            Some(&expected(include_str!(
                "../icon_resources/octicons/icons/alert-24.svg"
            )))
        );
    }

    #[cfg(feature = "bootstrap")]
    #[test]
    fn bootstrap_not_filled_should_work() {
        assert_eq!(
            bootstrap("alarm", false, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/bootstrap/icons/alarm.svg"
            )))
        );
    }

    #[cfg(feature = "bootstrap")]
    #[test]
    fn bootstrap_filled_should_work() {
        assert_eq!(
            bootstrap("alarm", true, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/bootstrap/icons/alarm-fill.svg"
            )))
        );
    }

    #[cfg(feature = "feather")]
    #[test]
    fn feather_should_work() {
        assert_eq!(
            feather("activity", IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/feather/icons/activity.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_brands_should_work() {
        assert_eq!(
            font_awesome("github", FontAwesome::Solid, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/font-awesome/svgs/brands/github.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_regular_should_work() {
        assert_eq!(
            font_awesome("address-book", FontAwesome::Regular, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/font-awesome/svgs/regular/address-book.svg"
            )))
        );
    }

    #[cfg(feature = "font-awesome")]
    #[test]
    fn font_awesome_solid_should_work() {
        assert_eq!(
            font_awesome("address-book", FontAwesome::Solid, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/font-awesome/svgs/solid/address-book.svg"
            )))
        );
    }

    #[cfg(feature = "heroicons")]
    #[test]
    fn heroicons_outline_should_work() {
        assert_eq!(
            heroicons("academic-cap", true, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/heroicons/optimized/24/outline/academic-cap.svg"
            )))
        );
    }

    #[cfg(feature = "heroicons")]
    #[test]
    fn heroicons_solid_should_work() {
        assert_eq!(
            heroicons("academic-cap", false, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/heroicons/optimized/24/solid/academic-cap.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_outline_should_work() {
        assert_eq!(
            ionicons("alarm", Ionicons::Outline, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/ionicons/src/svg/alarm-outline.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_sharp_should_work() {
        assert_eq!(
            ionicons("alarm", Ionicons::Sharp, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/ionicons/src/svg/alarm-sharp.svg"
            )))
        );
    }

    #[cfg(feature = "ionicons")]
    #[test]
    fn ionicons_should_work() {
        assert_eq!(
            ionicons("logo-github", Ionicons::Normal, IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/ionicons/src/svg/logo-github.svg"
            )))
        );
    }

    #[cfg(feature = "octicons")]
    #[test]
    fn octicons_should_work() {
        assert_eq!(
            octicons("alert", IconAttrs::default()),
            Some(expected(include_str!(
                "../icon_resources/octicons/icons/alert-24.svg"
            )))
        );
    }

    #[cfg(feature = "bootstrap")]
    #[test]
    fn bootstrap_with_class_should_work() {
        let attrs = IconAttrs::default()
            .class("h-8 w-8 text-white")
            .fill("none")
            .stroke_color("currentColor");

        let icon = bootstrap("alarm", false, attrs).expect("exists");
        assert_eq!(&icon[..32], "<svg class=\"h-8 w-8 text-white\" ");
    }

    #[cfg(all(feature = "heroicons", feature = "json"))]
    #[test]
    fn json_attribute_should_work() {
        let attrs = &serde_json::json!({
            "class": "h-8 w-8 text-white",
            "fill": "none",
            "stroke_color": "currentColor",
        });

        let icon = heroicons("academic-cap", true, attrs.into()).expect("exists");
        assert_eq!(&icon[..32], "<svg class=\"h-8 w-8 text-white\" ");
    }

    #[test]
    fn icon_should_not_exist() {
        assert_eq!(get(IconType::Feather(Feather::Normal), "not_exist"), None);
    }

    fn expected(s: &str) -> String {
        String::from_utf8(minify(s.as_bytes(), &CFG)).unwrap()
    }
}
